### PR TITLE
Added label feature for the techui.yaml

### PR DESCRIPTION
**Pull request by adedamola-sode**
The changes include adding fields labels and child labels to techui.yaml components, which get passed as into the component models and entities. Then from entities, we get generated screens that take into consideration the child labels for the names of the screens, and also the main label instead of a key of components in the yaml file. After the screens are generated, the generate_json_map uses a new function _get_labels to fetch the labels from the techui.yaml, and places them where needed, taking precedence before what is already in the screens.

### CODE REVIEW

File: src/techui_builder/generate.py [hunk 1]
```diff
@@ -173,6 +175,18 @@ def _initialise_name_suffix(self, component: Entity) -> tuple[str, str, str | No
             suffix = ""
             suffix_label = ""
 
+            # Try to get name from child labels if they exist,
```
Comment (OCopping) [inline src/techui_builder/generate.py]: Small gripe, but the indentation here is wrong 😅

File: src/techui_builder/generate.py [hunk 2]
```diff
@@ -173,6 +175,18 @@ def _initialise_name_suffix(self, component: Entity) -> tuple[str, str, str | No
...
+            # if not, just use the name as it is.
+        if component.child_labels is not None:
+            if (
+                name.removeprefix(":").removesuffix(":")
+                in component.child_labels.keys()
+            ):
+                name = component.child_labels[name.removeprefix(":").removesuffix(":")]
```
Comment (OCopping) [inline src/techui_builder/generate.py]: As this is reused, it would be tidier to put it in a temporary variable, e.g.

```suggestion
            clean_name = name.removeprefix(":").removesuffix(":")
            if raw_name in component.child_labels.keys():
                name = component.child_labels[clean_name]
```

File: src/techui_builder/generate.py [hunk 3]
```diff
@@ -381,8 +398,16 @@ def build_groups(self, screen_name: str):
         # that will be created.
         height, width = self._get_group_dimensions(self.widgets)
 
+        if (
+            screen_name in self.components
```
Comment (OCopping) [inline src/techui_builder/generate.py]:
```suggestion
            screen_name in self.components.keys()
```

File: tests/conftest.py [hunk 4]
```diff
@@ -109,10 +109,10 @@ def example_display_names_json():
 
 
 @pytest.fixture
-def generator():
+def generator(builder: Builder):
```
Comment (OCopping) [inline tests/conftest.py]: I would prefer we didn't pass the `Builder` fixture into the `Generator` fixture. Can you make a mock components dict instead? Or even make a fixture that returns a mock components list as I think this is used elsewhere too

File: tests/test_builder.py [hunk 5]
```diff
@@ -300,7 +304,7 @@ def test_generate_json_map(
     mock_get_action_group.return_value = mock_xml
 
     test_json_map = builder_with_test_files._generate_json_map(
-        screen_path.absolute(), dest_path
+        screen_path.absolute(), dest_path, builder_with_test_files.conf.components
```
Comment (OCopping) [inline tests/test_builder.py]: See https://github.com/DiamondLightSource/techui-builder/pull/216/changes#r3193774851

File: tests/test_builder.py [hunk 6]
```diff
@@ -311,7 +315,9 @@ def test_generate_json_map_embedded_screen(builder_with_test_files, example_json
     screen_path = Path("tests/test_files/test_bob_embedded.bob").absolute()
     dest_path = Path("tests/test_files/")
 
-    test_json_map = builder_with_test_files._generate_json_map(screen_path, dest_path)
+    test_json_map = builder_with_test_files._generate_json_map(
+        screen_path, dest_path, builder_with_test_files.conf.components
```
Comment (OCopping) [inline tests/test_builder.py]: See https://github.com/DiamondLightSource/techui-builder/pull/216/changes#r3193774851

File: tests/test_builder.py [hunk 7]
```diff
@@ -471,3 +483,35 @@ def test_get_action_group_no_actions_group(caplog):
...
+def test_get_labels_child_labels_with_name_already_pregenerated(
+    builder_with_test_files,
+):
+    display_name = _get_labels(
+        "X1",
+        builder_with_test_files.conf.components,
+        current_component_name="motor",
+        display_name="X",
+    )
+    assert display_name == "X1"
```
Comment (OCopping) [inline tests/test_builder.py]: These tests would need to be changed re- https://github.com/DiamondLightSource/techui-builder/pull/216/changes#r3193774851

File: tests/test_generate.py [hunk 8]
```diff
@@ -342,6 +359,42 @@ def test_generator_build_screen(generator):
...
+    generator.layout_widgets = Mock(
+        return_value=[
+            pwidget.EmbeddedDisplay(name="X", file="", x=0, y=0, width=205, height=120),
+            pwidget.EmbeddedDisplay(
+                name="Y", file="", x=0, y=150, width=205, height=120
+            ),
+        ]
+    )
+    generator._get_group_dimensions = Mock(return_value=(600, 400))
+    generator.build_groups(screen_name)
```
Comment (OCopping) [inline tests/test_generate.py]: Minor tidiness adjustment 😅 
```suggestion
    generator._get_group_dimensions = Mock(return_value=(600, 400))

    generator.build_groups(screen_name)
```

File: tests/test_generate.py [hunk 9]
```diff
@@ -342,6 +359,42 @@ def test_generator_build_screen(generator):
...
```
Comment (OCopping) [inline tests/test_generate.py]:
```suggestion
    generator._get_group_dimensions = Mock(return_value=(600, 400))

    generator.build_groups(screen_name)
```

File: src/techui_builder/builder.py [hunk 10]
```diff
@@ -499,3 +541,34 @@ def _get_action_group(element: ObjectifiedElement) -> ObjectifiedElement | None:
...
+    component: dict[str, Component],
+    current_component_name: str | None,
+    display_name: str | None,
+) -> str | None:
+    """
+    Get display name from child labels if they exist, otherwise return name_elem
+     or existing display_name if name_elem is None.
+    """
+    if name_elem is not None:
+        if name_elem in component.keys() and component[name_elem].label is not None:
```
Comment (OCopping) [inline src/techui_builder/builder.py]: Can you write a test for if `name_elem` exists but it isn't in `component.keys()`?

File: src/techui_builder/builder.py [hunk 11]
```diff
@@ -499,3 +541,34 @@ def _get_action_group(element: ObjectifiedElement) -> ObjectifiedElement | None:
...
+            display_name = component[name_elem].label
+        elif current_component_name is not None and (
```
Comment (OCopping) [inline src/techui_builder/builder.py]: I think the indentation here is wrong maybe? Also can you write a test for if `current_component_name` isn't `None` but it isn't in the `component` dict?

**Review by OCopping — CHANGES_REQUESTED** [review]

File: tests/conftest.py (hunk 4 above)
Comment (adedamola-sode) [reply tests/conftest.py]: Since I am trying to use the test-services techui-yaml, is it fine if I have the components fixture take in builder instead? Since that's how it's also done in the actual codebase?

File: src/techui_builder/generate.py
Comment (OCopping) [inline src/techui_builder/generate.py]: Instead of creating a new class var for self.components, why not do a similar thing to `build_wdiegst()` where `screen_components` is passed in from builder?
You would need to rework the logic below to work for a list, or would have to refactor the code in `create_screens()` in `builder.py` to be a dict

**Review by OCopping — CHANGES_REQUESTED** [review]

File: src/techui_builder/generate.py
Comment (adedamola-sode) [reply src/techui_builder/generate.py]: I passed in components as an argument this time instead

File: src/techui_builder/builder.py (hunk 11 above)
Comment (adedamola-sode) [reply src/techui_builder/builder.py]: I can't perceive anything wrong with the indentation as the linter says it's fine. I have written the test for current_component_name

File: src/techui_builder/builder.py (hunk 10 above)
Comment (adedamola-sode) [reply src/techui_builder/builder.py]: Done

File: src/techui_builder/generate.py (hunk 1 above)
Comment (adedamola-sode) [reply src/techui_builder/generate.py]: I can't tell what's wrong with the indentation, especially with no correction coming from ruff

File: tests/test_builder.py
Comment (OCopping) [inline tests/test_builder.py:406]: You will need to patch `_get_labels()` out of all tests that call `_generate_json_map()` to make sure that code isn't called in these tests.

**Review by OCopping — CHANGES_REQUESTED** [review]

File: tests/test_builder.py [hunk 12]
```diff
@@ -306,6 +308,7 @@ def test_generate_json_map(
     mock_xml = objectify.Element("action")
     mock_xml["file"] = "test_child_bob.bob"
     mock_get_action_group.return_value = mock_xml
+    mock_get_labels.side_effect = ["Display", "Detector"]
```
Comment (OCopping) [inline tests/test_builder.py]: This should be `return_value` not `side_effect`

File: tests/test_builder.py [hunk 13]
```diff
@@ -315,21 +318,26 @@ def test_generate_json_map(
 
 
 # TODO: write this test
+@patch("techui_builder.builder._get_labels")
 def test_generate_json_map_embedded_screen(
-    builder_with_test_files, example_json_map, components
+    mock_get_labels, builder_with_test_files, example_json_map, components
 ):
+    mock_get_labels.side_effect = ["Display", "Detector", "Embedded Display"]
```
Comment (OCopping) [inline tests/test_builder.py]: This should be `return_value` not `side_effect`

File: tests/test_builder.py [hunk 14]
```diff
@@ -412,12 +422,12 @@ def test_generate_json_map_get_macros(
     macros = objectify.SubElement(mock_xml, "macros")
     # Set a macro to test
     macros["macro"] = "value"
+    mock_get_labels.side_effect = ["Display", "Detector"]
```
Comment (OCopping) [inline tests/test_builder.py]: This should be `return_value` not `side_effect`

File: tests/test_builder.py [hunk 15]
```diff
@@ -435,12 +445,18 @@ def test_generate_json_map_xml_parse_error(
...
+    mock_get_labels,
+    mock_get_action_group,
+    builder_with_test_files,
+    test_files,
+    components,
 ):
     screen_path, dest_path = test_files
 
     mock_get_action_group.side_effect = Exception("Some exception")
+    mock_get_labels.side_effect = ["Display", "Detector"]
```
Comment (OCopping) [inline tests/test_builder.py]: This should be `return_value` not `side_effect`

**Review by OCopping — CHANGES_REQUESTED** [review]

File: src/techui_builder/builder.py [hunk 16]
```diff
@@ -499,3 +539,36 @@ def _get_action_group(element: ObjectifiedElement) -> ObjectifiedElement | None:
             f"Actions group not found in component [bold]{name}[/bold] on "
             f"[bold]{parent_name}[/bold]"
         )
+
+
+def _get_labels(
```
Comment (OCopping) [inline src/techui_builder/builder.py]: Rename this to `get_label` so that it is clear it only returns one string and not a collection of strings

```suggestion
def _get_label(
```

**Review by OCopping — CHANGES_REQUESTED** [review]

File: src/techui_builder/builder.py [hunk 17]
```diff
@@ -251,7 +257,14 @@ def create_screens(self):
                     " any P field in the ioc.yaml files in services"
                 )
 
-    def _generate_json_map(self, screen_path: Path, dest_path: Path) -> JsonMap:
+    def _generate_json_map(
+        self,
+        screen_path: Path,
+        dest_path: Path,
+        component: dict[str, Component],
```
Comment (OCopping) [inline src/techui_builder/builder.py]: I'm not too fond of this, as it means you are passing the same dictionary through every level of the recursion instead of just the component we want at the time.
I'm not sure what a cleaner way to do this would be.

Maybe instead of passing `component` in to the function, can we just refer to `self.conf.components` inside the function?
Even his I'm not sure of, as then we are relying on a variable from outside the scope of the function... It might be fine though...

**Review by OCopping — CHANGES_REQUESTED** [review]

**Review by OCopping — APPROVED** [review]
